File: ch06/VB/Data/Data/DataForm.vb
Public Class DataForm
    Private _source As New ObjectSource()

    Private Sub DataForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        CategoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName"
        CategoryToolStripComboBox.ComboBox.ValueMember = "CategoryID"
        CategoryToolStripComboBox.ComboBox.DataSource = _source.GetCategories()
    End Sub

    Private Sub CategoryToolStripComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles CategoryToolStripComboBox.SelectedIndexChanged
        Dim catId = CategoryToolStripComboBox.ComboBox.SelectedValue
        Dim products = _source.GetProducts(catId)

        ProductsListBox.DataSource = products
        ProductsListBox.DisplayMember = "ProductName"
        ProductsDataGridView.DataSource = products

        NameTextBox.DataBindings.Clear()
        NameTextBox.DataBindings.Add("Text", products, "ProductName")
        QuantityPerUnitTextBox.DataBindings.Clear()
        QuantityPerUnitTextBox.DataBindings.Add("Text", products, "QuantityPerUnit")
        PriceTextBox.DataBindings.Clear()
        PriceTextBox.DataBindings.Add("Text", products, "UnitPrice")
        StockTextBox.DataBindings.Clear()
        StockTextBox.DataBindings.Add("Text", products, "UnitsInStock")
        UnitsOnOrderTextBox.DataBindings.Clear()
        UnitsOnOrderTextBox.DataBindings.Add("Text", products, "UnitsOnOrder")
        DiscontinuedCheckBox.DataBindings.Clear()
        DiscontinuedCheckBox.DataBindings.Add("Checked", products, "Discontinued")
    End Sub

    Private Sub AddToolStripButton_Click(sender As Object, e As EventArgs) Handles AddToolStripButton.Click
        Dim category = CType(CategoryToolStripComboBox.SelectedItem, Category)
        Dim form As New AddProductForm(category)
        Dim result = form.ShowDialog()
        If result = DialogResult.OK Then
            _source.AddProduct(form.Product)
        End If

    End Sub

    Private Sub DeleteToolStripButton_Click(sender As Object, e As EventArgs) Handles DeleteToolStripButton.Click
        Dim product = CType(ProductsListBox.SelectedItem, Product)
        _source.DeleteProduct(product)
    End Sub
End Class

File: ch06/CS/Data/Data/DataForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Data
{
    public partial class DataForm : Form
    {
        public DataForm()
        {
            InitializeComponent();
        }

        private ObjectSource source = new ObjectSource();
        private BindingSource categoryBindingSource = new BindingSource();
        private BindingSource productsBindingSource = new BindingSource();

        private void DataForm_Load(object sender, EventArgs e)
        {
            categoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName";
            categoryToolStripComboBox.ComboBox.ValueMember = "CategoryID";
            categoryToolStripComboBox.ComboBox.DataSource = source.GetCategories();

            categoryBindingSource.DataSource = source.GetCategories();

            productsListBox.DataSource = productsBindingSource;
            productsListBox.DisplayMember = "ProductName";

            nameTextBox.DataBindings.Add("Text", productsBindingSource, "ProductName");
            quantityPerUnitTextBox.DataBindings.Add("Text", productsBindingSource, "QuantityPerUnit");
            priceTextBox.DataBindings.Add("Text", productsBindingSource, "UnitPrice");
            stockTextBox.DataBindings.Add("Text", productsBindingSource, "UnitsInStock");
            orderTextBox.DataBindings.Add("Text", productsBindingSource, "UnitsOnOrder");
            discontinuedCheckBox.DataBindings.Add("Checked", productsBindingSource, "Discontinued");
        }

        private void categoryToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            var catId = Convert.ToInt32(categoryToolStripComboBox.ComboBox.SelectedValue);
            productsBindingSource.DataSource = source.GetProducts(catId);
        }

        private void AddToolStripButton_Click(object sender, EventArgs e)
        {
            var form = new AddProductForm((Category)categoryToolStripComboBox.ComboBox.SelectedItem);
            var result = form.ShowDialog();
            if (result == DialogResult.OK)
            {
                //source.AddProduct(form.Product);
                productsBindingSource.Add(form.Product);
            }
        }

        private void DeleteToolStripButton_Click(object sender, EventArgs e)
        {
            //source.DeleteProduct((Product)productsListBox.SelectedItem);
            productsBindingSource.Remove(productsListBox.SelectedItem);
        }

        private void BackToolStripButton_Click(object sender, EventArgs e)
        {
            productsBindingSource.MovePrevious();
        }

        private void ForwardToolStripButton_Click(object sender, EventArgs e)
        {
            productsBindingSource.MoveNext();
        }
    }
}
